avgsamplerate.go

```diff
@@ -172,3 +176,38 @@ func (a *AvgSampleRate) GetSampleRate(key string) int {
 	}
 	return 1
 }
+
+type avgSampleRateState struct {
+	SavedSampleRates map[string]int `json:"saved_sample_rates"`
+}
+
+func (a *AvgSampleRate) SaveState() ([]byte, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if a.savedSampleRates == nil {
+		return nil, errors.New("saved sample rate map is nil")
+	}
+	s := &avgSampleRateState{SavedSampleRates: a.savedSampleRates}
+	return json.Marshal(s)
+}
+
+func (a *AvgSampleRate) LoadState(state []byte) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	s := avgSampleRateState{}
+	err := json.Unmarshal(state, &s)
+	if err != nil {
+		return err
+	}
+
+	// Load the previously calculated sample rates
+	a.savedSampleRates = s.SavedSampleRates
+	// Prepare for counting new keys
+	a.currentCounts = make(map[string]int)
```

Comment:
why is `currentCounts` initialized here instead of in `Start`? If it weren't initialized here, it could be taken out of the conditional in `Start`.